### PR TITLE
Avoid mutating argument when expanding parameters

### PR DESCRIPTION
**Pull request by jadefish**
`expandParams` mutates the object argument passed in by the action
caller (`VPosts`, `VReplaces`, `VLoads`). For action parameters (e.g.,
via `last_response`), this results in the action parameter's structure
(which indicates to `VActionParameter` how to resolve the value) being
overwritten with the resolved value the first time the action's
parameters are expanded (the first time the action is invoked). As a
result, the first resolved value for the parameter is persisted for all
subsequent invocations of the action, and the parameter is never again
expanded.

This change makes `expandParams` collect resolved parameters into a new
object, avoiding the mutation of the object argument passed in by the
action. The initial
structure of all action parameters remains intact, and subsequent
invocations of the action can continue to expand action parameters as
expected.

### CODE REVIEW

File: views/mdc/assets/js/components/events/action_parameter.js [hunk 1]
```diff
@@ -25,15 +25,21 @@ export function expandParam(results, value) {
...
-export function expandParams(results, o) {
-    Object.keys(o).forEach(function(k) {
-        if (o[k] !== null && typeof o[k] === 'object' && o[k].type !==
-            'action_parameter') {
-            expandParams(results, o[k]);
-            return;
+export function expandParams(results, params) {
+    const expandedParams = {};
+
+    for (const [key, value] of Object.entries(params)) {
```
Comment (NullTerminator) [inline views/mdc/assets/js/components/events/action_parameter.js]: This comment isn't exactly for this line, but here we go:  This is only assigning values to `expandedParams` if they are an attribute of `params` that is an object.  So if params is `{ foo: 4 }` this'll result in `{}`.  I'm not entirely sure of the data flowing through this function though.  If we only care about things that are action parameters then we're good, but if we want everything it's going to have to start as a deep copy of `params` and then expand itself.

File: views/mdc/assets/js/components/events/action_parameter.js [hunk 2]
```diff
@@ -25,15 +25,21 @@ export function expandParam(results, value) {
...
+        if (typeof value !== 'object') {
```
[the remark this answers is not in the document]
Comment (jadefish) [reply views/mdc/assets/js/components/events/action_parameter.js]: Good catch. `typeof null === 'object'` will forever "gotcha" me.

File: views/mdc/assets/js/components/events/action_parameter.js (hunk 1 above)
Comment (jadefish) [reply views/mdc/assets/js/components/events/action_parameter.js]: I wasn't 100% sure about this, so I dug into it a little. It seems that only action parameters make it into this function, and simple values are built into the URL.

```ruby
replaces :something, :something_else,
         foo: 'bar',
         key: last_response.data
```

generates `GET /something_else?foo=bar` and passes `{"key":{"value":["data"],"type":"action_parameter","response_index":0}}` to `expandParams`.

I think is is coincidental, and if invoked independently, the previous version would indeed preserve non-`action_parameter` values.

Comment (jadefish) [reply views/mdc/assets/js/components/events/action_parameter.js]: I've pushed a change to preserve non-`action_parameter` values so `expandParams` will always behave consistently.

**Comment by rx** [comment]
I'd like to pull this down and take a look. Looks good when reading the code.
Adding this case to the demo pom would help.

**Pull request merged**